### ISSUE TITLE
Running Nanosaur on the Raspberry Pi 4

### ISSUE THREAD

**Issue by vanfanel** [jorio/Nanosaur#21]
Hi there,

Now that it's possible to build Nanosaur without GLU or X11 (once the pending PR is merged at least!), it's time to get it running on a Raspberry Pi 4 using the latest MESA OpenGL implementation.

Currently, I am getting this:

```
~/src/Nanosaur/b4# ./Nanosaur 
NANOSAUR FATAL ALERT: OpenGL error 0x500
in __SetInitialClientState:170
```

Seems to me that Nanosaur requires OpenGL 2.0, and the Pi4 does up to OpenGL 2.1 if I am not mistaken, so.. Any idea on what's missing here?

**Comment by vanfanel** [comment]
To actively contribute to a fix for this, I ran Nanosaur Pi4 build on GDB, and found this:
```

Thread 1 "Nanosaur" hit Breakpoint 1, __SetInitialClientState (stateEnum=32884, stateFlagPtr=0x5555682035 <gState+5>, 
    initialValue=true) at /root/src/Nanosaur/src/QD3D/Renderer.c:165
165             *stateFlagPtr = initialValue;
(gdb) n
ALSA lib pcm.c:8545:(snd_pcm_recover) underrun occurred
166             if (initialValue)
(gdb) n
ALSA lib pcm.c:8545:(snd_pcm_recover) underrun occurred
167                     glEnableClientState(stateEnum);
(gdb) 
ALSA lib pcm.c:8545:(snd_pcm_recover) underrun occurred
170             CHECK_GL_ERROR();
(gdb)
```
As you can see, it is `stateEnum=32884`, which translates to:

`usr/include/GL/gl.h:#define GL_VERTEX_ARRAY                            0x8074`

...But I have no idea on why setting `GL_VERTEX_ARRAY` state to `true` is a problem on the Pi4.
I have some full OpenGL programs working on the Pi4 with no problems so in general it works.

**Comment by jorio** [comment]
Yeah, GL_VERTEX_ARRAY is the very first state that gets initialized in Render_InitState, so something may have errored out before entering Render_InitState.

I suppose some GL call went wrong in [QD3D_SetupWindow](https://github.com/jorio/Nanosaur/blob/master/src/QD3D/QD3D_Support.c#L121) between the creation of the GL context (which appears to be successful since we get this far) and the call to Render_InitState.

You can try adding some `CHECK_GL_ERROR();` lines after the GL calls in QD3D_SetupWindow to narrow down the culprit.

**Comment by vanfanel** [comment]
@jorio I am using `glError()` and I have found out that it returns 0 untill this exact call:

https://github.com/jorio/Nanosaur/blob/b567a3e6d7fd1cbc43800cfaa1bd82f31c6d9fae/src/Screens/Title.c#L342

Does it make any sense to you?
Only happens on the Pi, on several X86 machines I have around it works perfectly well.

**Comment by orazioedoardo** [comment]
@jorio I'm getting this same error, GL_INVALID_ENUM, on a clean installed Debian 13 virtual machine with GNOME, graphic `virgl(ANGLE(Apple, Apple M1, OpenGL 4.1 - Metal 89,4))`. The only game that works is Bugdom 1 since CHECK_GL_ERROR doesn't do anything on release mode. If i uncomment the assertion, all games seems to work while trying for a bit though.

**Comment by orazioedoardo** [comment]
Same on Ubuntu 26.04

**Comment by jorio** [comment]
I think it's a quirk with your GL driver. If `glEnableClientState(GL_VERTEX_ARRAY)` causes GL_INVALID_ENUM, that's your GL driver saying it doesn't support vertex arrays, which the game really needs. I don't think there's much to do about this beyond silencing the GL asserts in the game at the risk of missing other errors elsewhere.

I tried to reproduce the error you're seeing with virgl. I compiled the game from scratch from the latest commit on master (after installing [the dependencies for SDL3](https://github.com/libsdl-org/SDL/blob/main/docs/README-linux.md#build-dependencies)), and was able to run it without any errors on the following qemu guests:
- Ubuntu 24.04.4 x86_64 (Linux x86_64 host), "virgl (Mesa Intel(R) HD Graphics...)"
- Ubuntu 24.04.4 ARM64 (macOS ARM64 host), "virgl (ANGLE (Apple, ... OpenGL 4.1 Metal - 89.4))"

**Comment by orazioedoardo** [comment]
> I tried to reproduce the error you're seeing with virgl. I compiled the game from scratch from the latest commit on master (after installing [the dependencies for SDL3](https://github.com/libsdl-org/SDL/blob/main/docs/README-linux.md#build-dependencies))

I also tried compiling from the latest commit on ubuntu 26.04 (qemu guest on macos arm64) and I get the same error, default desktop install with the addition of [libopengl0](https://packages.ubuntu.com/resolute/libopengl0).

**Comment by jorio** [comment]
OK, I'm able to repro on a fresh install of a 26.04 in an ARM64 guest on a Mac host. The SDL_GL_CreateContext call in QD3D_Boot is responsible for the GL error - not vertex arrays (as another comment did mention a couple years ago, sorry for missing that in my previous reply).

I guess we could flush the error by calling glGetError() right after SDL_GL_CreateContext, but I still wonder what's causing this.

**Comment by jorio** [comment]
Please see the newest commit on master and let me know if that's enough to get the game running on your setup.

**Comment by orazioedoardo** [comment]
Nanosaur worked fine on a clean install of Ubuntu 26.04, though as above the AppImage requires libipengl0
